### PR TITLE
🌱 Set provider ID through kubelet in ubuntu templates

### LINKED ISSUE #1547: Node/server name mismatch due to dots in name
/kind bug

**What steps did you take and what happened:**

I created an OpenStackMachineTemplate with a name like `my-cluster-md-0-v1.26.4` and tried to use it.
This caused a strange issue where the kubelet of the new Node managed to register the Node, but the cloud controller then deleted the Node because it could not find a matching server.

Why is this? The name of the OpenStackMachine is derived from the OpenStackMachineTemplate and contains dots.
When translated to a host name the dots are translated to dashes somewhere (not sure where). Thus, the kubelet will register a name with dashes but the server in OpenStack has a name with dots and the cloud controller will get confused.

```console
❯ openstack server list
+--------------------------------------+----------------------------------+--------+-----------------------------------------------------------------------+---------------------------------------+--------------+
| ID                                   | Name                             | Status | Networks                                                              | Image                                 | Flavor       |
+--------------------------------------+----------------------------------+--------+-----------------------------------------------------------------------+---------------------------------------+--------------+
| cee32b6e-b1f8-4f41-97e8-07fb4ee73189 | lennart-test-md-0-v1.26.3-2rgwx  | ACTIVE | k8s-clusterapi-cluster-default-lennart-test=10.6.0.130                | ubuntu-2204-kube-v1.26.3              | 4C-4GB-100GB |
| c5732758-f16e-483c-a676-08685cd2cf12 | lennart-test-md-0-mphbw          | ACTIVE | k8s-clusterapi-cluster-default-lennart-test=10.6.0.99                 | ubuntu-2204-kube-v1.26.3              | 4C-4GB-100GB |
| 24df95e5-c823-4bc4-add0-a6952f5b1c77 | lennart-test-control-plane-nf4fh | ACTIVE | k8s-clusterapi-cluster-default-lennart-test=10.6.0.133                | ubuntu-2204-kube-v1.26.3              | 4C-4GB-100GB |
| 46f461f4-34b2-48cc-997f-8c6eebcc9b56 | lennart-test-bastion             | ACTIVE | k8s-clusterapi-cluster-default-lennart-test=10.6.0.243, <redacted> | Ubuntu 22.04.1 Jammy Jellyfish 230124 | 1C-2GB-50GB  |
+--------------------------------------+----------------------------------+--------+-----------------------------------------------------------------------+---------------------------------------+--------------+

❯ k get nodes
NAME                               STATUS   ROLES           AGE   VERSION
lennart-test-control-plane-nf4fh   Ready    control-plane   37m   v1.26.3
lennart-test-md-0-mphbw            Ready    <none>          30m   v1.26.3

❯ k get machine
NAME                                       CLUSTER        NODENAME                           PROVIDERID                                          PHASE         AGE     VERSION
lennart-test-control-plane-kbzhl           lennart-test   lennart-test-control-plane-nf4fh   openstack:///24df95e5-c823-4bc4-add0-a6952f5b1c77   Running       36m     v1.26.3
lennart-test-md-0-5fd584f555xwjd5q-9w6fk   lennart-test   lennart-test-md-0-mphbw            openstack:///c5732758-f16e-483c-a676-08685cd2cf12   Running       39m     v1.26.3
lennart-test-md-0-69457bc9b8xw8xjn-pl6vg   lennart-test                                      openstack:///cee32b6e-b1f8-4f41-97e8-07fb4ee73189   Provisioned   6m25s   v1.26.3

❯ k get openstackmachine
NAME                               CLUSTER        INSTANCESTATE   READY   PROVIDERID                                          MACHINE                                    AGE
lennart-test-control-plane-nf4fh   lennart-test   ACTIVE          true    openstack:///24df95e5-c823-4bc4-add0-a6952f5b1c77   lennart-test-control-plane-kbzhl           37m
lennart-test-md-0-mphbw            lennart-test   ACTIVE          true    openstack:///c5732758-f16e-483c-a676-08685cd2cf12   lennart-test-md-0-5fd584f555xwjd5q-9w6fk   39m
lennart-test-md-0-v1.26.3-2rgwx    lennart-test   ACTIVE          true    openstack:///cee32b6e-b1f8-4f41-97e8-07fb4ee73189   lennart-test-md-0-69457bc9b8xw8xjn-pl6vg   6m37s
```

Cloud controller logs:

```
I0512 10:01:03.068981       1 openstack.go:359] Claiming to support Zones
I0512 10:01:03.315683       1 node_controller.go:484] Successfully initialized node lennart-test-control-plane-nf4fh with cloud provider
I0512 10:01:03.315732       1 node_controller.go:415] Initializing node lennart-test-md-0-mphbw with cloud provider
I0512 10:01:03.316150       1 event.go:294] "Event occurred" object="lennart-test-control-plane-nf4fh" fieldPath="" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
I0512 10:01:04.439483       1 openstack.go:359] Claiming to support Zones
I0512 10:01:04.701816       1 node_controller.go:484] Successfully initialized node lennart-test-md-0-mphbw with cloud provider
I0512 10:01:04.702219       1 event.go:294] "Event occurred" object="lennart-test-md-0-mphbw" fieldPath="" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
I0512 10:10:25.179835       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:25.339485       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:25.339517       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:25.487962       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:25.488143       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:25.654573       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:25.654614       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:25.806657       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:25.806761       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:25.964009       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:25.964049       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:26.133396       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:26.133431       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
E0512 10:10:26.269076       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
I0512 10:10:26.294300       1 node_controller.go:415] Initializing node lennart-test-md-0-v1-26-3-2rgwx with cloud provider
I0512 10:10:26.450831       1 event.go:294] "Event occurred" object="lennart-test-md-0-v1-26-3-2rgwx" fieldPath="" kind="Node" apiVersion="" type="Normal" reason="DeletingNode" message="Deleting node lennart-test-md-0-v1-26-3-2rgwx because it does not exist in the cloud provider"
E0512 10:10:26.472366       1 node_controller.go:229] error syncing 'lennart-test-md-0-v1-26-3-2rgwx': failed to get provider ID for node lennart-test-md-0-v1-26-3-2rgwx at cloudprovider: failed to get instance ID from cloud provider: instance not found, requeuing
```

**What did you expect to happen:**

I would expect that this is handled either by

1. rejecting invalid names at some level (webhook when creating OpenStackMachineTemplate?)
2. handle it transparenlty, e.g. by transforming the name in the same way everywhere so that there is no mismatch, or by relying on a different way to match the servers to the nodes (providerID?).

**Anything else you would like to add:**

Initially reported in CAPI https://github.com/kubernetes-sigs/cluster-api/issues/8631.

Just want to highlight also how dangerous this bug is when happening with control plane nodes. The kubelet is able to register the node and the etcd instance is able to become a member. But CAPI/CAPO is unable to remove this etcd member later since the node doesn't exist. This means that it is very easy to lose quorum and break the cluster! :boom: 

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): v0.7.2
- Cluster-API version: v1.4.2
- OpenStack version: Victoria
- Minikube/KIND version:
- Kubernetes version (use `kubectl version`): v1.26.4
- OS (e.g. from `/etc/os-release`): ubuntu 22.04

### PR DESCRIPTION
**Pull request by lentzi90**
**What this PR does / why we need it**:

To avoid issues with potential mismatches between node names and
openstack servers we can configure the kubelet to set the provider ID.
Otherwise, the cloud controller will try to match nodes and servers
based on just the name. The names can differ because of special
characters, like dots. When this happens, the cloud controller will be
unable to match them and thus believe that the node has no underlying
server.

Also removes some unused (I think) cluster templates and kustomizations.

I will make a follow up PR for the flatcar templates.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1547 or at least works around it.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
2. I tagged this as :seedling: not sure if it makes sense

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Review by mdbooth — COMMENTED** [review]
/lgtm

**Comment by jichenjc** [comment]
I am curious on `{{ instance_id }}` comes from..  is this a Cluster API definition and CAPO feed for it somewhere? 
I didn't see it used currently in CAPO now ..so curious how the template become real provider ID..

**Comment by lentzi90** [comment]
> I am curious on `{{ instance_id }}` comes from.. is this a Cluster API definition and CAPO feed for it somewhere? I didn't see it used currently in CAPO now ..so curious how the template become real provider ID..

It comes with the [instance data from cloud-init](https://cloudinit.readthedocs.io/en/latest/explanation/instancedata.html). Basically all of the kubeadm config will be injected through cloud-init (or ignition) and can have templated values that is filled in from the instance data.

**Comment by lentzi90** [comment]
Hmm I'm trying to figure out where the issue is for flatcar but not finding any relevant logs. Any clue @tormath1 ?

**Comment by mdbooth** [comment]
There are 2 problems here:
1. coreos-metadata (aka afterburn?) is using OpenStack's EC2 metadata service rather than OpenStack's native metadata service.
2. It seems there's a bug in ec2 metadata which means it's returning i-<libvirt instance name> rather than the actual uuid.

And a third problem:
3. ec2 metadata is deep frozen, hopelessly bitrotted, and chances are nobody's going to fix it.

We should probably try to fix it in afterburn, tbh. @tormath1 what's the relationship between afterburn and coreos-metadata?

**Comment by mdbooth** [comment]
> I can open an issue in afterburn. (From what I can see the coreos-metadata is the same as afterburn.) For some reason they are checking the EC2 service when [fetching from network](https://github.com/coreos/afterburn/blob/f22131c775094c559baf5c5d712425f341f3556e/src/providers/openstack/network.rs#L11). When using config-drive it combines some fields from the EC2 service and some from OpenStack, which seems really weird to me.
> 
> https://github.com/coreos/afterburn/blob/f22131c775094c559baf5c5d712425f341f3556e/src/providers/openstack/configdrive.rs#L146-L151

Beat you to it: https://github.com/coreos/afterburn/issues/930

**Comment by mdbooth** [comment]
Because they're separate templates, we have the option to fix this in Flatcar independently right? I wonder if we move on with the Ubuntu fix for now and raise a separate issue to fix it for Flatcar. @lentzi90 @tormath1 Thoughts?

**Comment by lentzi90** [comment]
Yes that works for me! I can split this PR into one for ubuntu and one for flatcar.

File: kustomize/v1alpha7/flatcar/patch-flatcar.yaml [hunk 1]
```diff
@@ -17,9 +17,13 @@ spec:
     joinConfiguration:
       nodeRegistration:
         name: $${COREOS_OPENSTACK_HOSTNAME}
+        kubeletExtraArgs:
+          provider-id: null
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply kustomize/v1alpha7/flatcar/patch-flatcar.yaml:21]: Can we add a comment with a link to the CAPO issue?

[the remark this answers is not in the document]
Comment (lentzi90) [reply kustomize/v1alpha7/flatcar/patch-flatcar.yaml:21]: Done :heavy_check_mark:

File: kustomize/v1alpha7/flatcar/patch-flatcar.yaml [hunk 2]
```diff
@@ -17,9 +17,16 @@ spec:
...
     initConfiguration:
       nodeRegistration:
         name: $${COREOS_OPENSTACK_HOSTNAME}
+        kubeletExtraArgs:
+          # Fixme(lentzi90): This is here just to override the value set in the default
+          # kustomization. It will be replaced with a value that works for flatcar in
+          # https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1564
```
Comment (mdbooth) [inline kustomize/v1alpha7/flatcar/patch-flatcar.yaml:28]: You added this comment in both places other than the one where I asked for it 🤣 I'll try not to take it personally.

It's not worth updating imho. The important thing is that that somebody looking will see the comment, and they'll see one of these.

File: templates/cluster-template-without-lb.yaml [hunk 3]
```diff
@@ -20,6 +20,7 @@ spec:
         nodeRegistration:
           kubeletExtraArgs:
             cloud-provider: external
+            provider-id: openstack:///'{{ instance_id }}'
```
Comment (mdbooth) [inline templates/cluster-template-without-lb.yaml:23]: Any idea why we're not inheriting this from default? The answer to this question would be outside the scope of this PR, though.

**Review by mdbooth — COMMENTED** [review]
/lgtm

File: kustomize/v1alpha7/flatcar/patch-flatcar.yaml (hunk 2 above)
Comment (lentzi90) [reply kustomize/v1alpha7/flatcar/patch-flatcar.yaml:28]: OMG sorry! I'm not sure how I managed to miss the first occurance :facepalm: :joy:

File: templates/cluster-template-without-lb.yaml (hunk 3 above)
Comment (lentzi90) [reply templates/cluster-template-without-lb.yaml:23]: This is the generated template! So it is actually inherited from default and that is why we see it here. :slightly_smiling_face: The without-lb patch is [over here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/kustomize/v1alpha7/without-lb/patch-without-lb.yaml) and does not have any provider-id set.

**Pull request merged**